### PR TITLE
Aggregate ops memory safety

### PR DESCRIPTION
**Pull request by nano-o**
This depends on `conditional_points_to_untyped`

### CODE REVIEW

File: proof/aggregate.saw [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+*/
+
+///////////////////////////////////////////////////////////////////////////////
+// Specifications
+///////////////////////////////////////////////////////////////////////////////
+
+let blst_pairing_init_spec DST_len = do {
+  ctx_ptr <- crucible_alloc (llvm_struct "struct.PAIRING");
+  hash_or_encode <- crucible_fresh_var "hash_or_encode" (llvm_int 32);
+  //(DST, DST_ptr) <- ptr_to_fresh_readonly "DST" (llvm_array DST_len (llvm_int 8)); // not strictly needed for this function
```
Comment (andreistefanescu) [inline proof/aggregate.saw:13]: is `DST_ptr` not used?

File: proof/aggregate.saw [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+  crucible_points_to (crucible_field ctx_ptr "DST_len") (crucible_term {{`DST_len:[64]}});
+};
+
+// the enum for aggregate-operations errors
+let AGGR_UNDEFINED = {{ zero:[32] }};
+let AGGR_MIN_SIG = {{ (zext 0x1):[32] }};
+let AGGR_MIN_PK = {{ (zext 0x2):[32] }};
+let AGGR_SIGN_SET = {{ (zext 0x10):[32] }};
+let AGGR_GT_SET = {{ (zext 0x20):[32] }};
+let AGGR_HASH_OR_ENCODE = {{ (zext 0x40):[32] }};
```
Comment (andreistefanescu) [inline proof/aggregate.saw:29]: I don't think `zext` is needed

File: proof/aggregate.saw [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+
+// BLST_ERROR enum:
+let BLST_SUCCESS = 0;
+let BLST_AGGR_TYPE_MISMATCH = 4;
+let BLST_PK_IS_INFINITY = 6;
+let BLST_POINT_NOT_IN_GROUP = 3;
+
+let N_MAX = 8; // NOTE: must match N_MAX defined in C code
+
+let blst_pairing_aggregate_pk_in_g1_spec DST_len msg_len aug_len nelems null_sig = do {
```
Comment (andreistefanescu) [inline proof/aggregate.saw:40]: can this be combined with `blst_pairing_aggregate_pk_in_g2_spec`?

**Review by andreistefanescu — APPROVED** [review]
lgtm!

File: proof/aggregate.saw (hunk 1 above)
Comment (nano-o) [reply proof/aggregate.saw:13]: The function copies the pointer to `ctx->DST`, but it does not read what's pointed to.

File: proof/aggregate.saw (hunk 2 above)
Comment (nano-o) [reply proof/aggregate.saw:29]: It fails with a type mismatch without `zext`. I can use decimal notation to get rid of `zext` though.

File: proof/aggregate.saw (hunk 3 above)
Comment (nano-o) [reply proof/aggregate.saw:40]: Yes it probably should!

File: proof/aggregate.saw [hunk 4]
```diff
@@ -37,81 +36,38 @@ let BLST_POINT_NOT_IN_GROUP = 3;
...
+let POINTonE2_in_G2_spec = do {
+  (p, p_ptr) <- ptr_to_fresh_readonly "p" (llvm_struct "struct.POINTonE2_affine");
+  crucible_execute_func [p_ptr];
+  ret <- crucible_fresh_var "ret" limb_type;
+  crucible_postcond {{ ret == is_in_G2 p }};
   crucible_return (crucible_term ret);
 };
 
-let blst_pairing_aggregate_pk_in_g2_spec DST_len msg_len aug_len nelems null_sig = do {
+let blst_pairing_aggregate_pk_in_spec DST_len msg_len aug_len nelems null_sig min_sig = do {
```
Comment (andreistefanescu) [inline proof/aggregate.saw:57]: nice :-)

**Pull request merged**